From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 13 Sep 2024 06:10:38 +0300
Subject: [PATCH] Handle special nbt cases for Item Modifiers API

The following cases are currently handled:
- Items in bundles
- Items in containers (shulkers, chests, etc.)
- Leftover item in food component

diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
index dac68ae91a239cfcf5866b28d1a5654a7485f8a4..60a7004555e9e9ce5192e544b6b0be16e5a514da 100644
--- a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
@@ -125,6 +125,54 @@ public class ItemModifiersHandlerImpl extends ItemModifiersHandler {
             modified = true;
         }
 
+        // Special nbt cases
+        if (!contextBox.getItem().getType().isAir()) {
+            // Bundles
+            if (contextBox.getMeta() instanceof org.bukkit.inventory.meta.BundleMeta meta && meta.hasItems()) {
+                List<org.bukkit.inventory.ItemStack> items = new ArrayList<>(meta.getItems());
+                for (int i = 0; i < items.size(); i++) {
+                    org.bukkit.inventory.ItemStack item = items.get(i);
+                    org.bukkit.inventory.ItemStack newItem = ItemModifier.modifyItem(new ItemContextBox(contextBox.getViewer(), contextBox.getLocale(), ItemModifierContextType.EMPTY_NO_LORE, ItemModifierContext.EMPTY, item));
+                    if (newItem != null) {
+                        items.set(i, item);
+                        modified = true;
+                    }
+                }
+                if (modified) meta.setItems(items);
+            }
+            // Block storages
+            if (contextBox.getMeta() instanceof org.bukkit.inventory.meta.BlockStateMeta meta && meta.hasBlockState()) {
+                org.bukkit.block.BlockState blockState = meta.getBlockState();
+                if (blockState instanceof org.bukkit.block.Container container) {
+                    var inventory = container.getInventory();
+                    for (int i = 0; i < inventory.getSize(); i++) {
+                        org.bukkit.inventory.ItemStack item = inventory.getItem(i);
+                        if (item == null) continue;
+
+                        org.bukkit.inventory.ItemStack newItem = ItemModifier.modifyItem(new ItemContextBox(contextBox.getViewer(), contextBox.getLocale(), ItemModifierContextType.EMPTY_NO_LORE, ItemModifierContext.EMPTY, item));
+                        if (newItem != null) {
+                            inventory.setItem(i, item);
+                            modified = true;
+                        }
+                    }
+                    if (modified) meta.setBlockState(blockState);
+                }
+            }
+            // Food leftover
+            if (contextBox.getMeta().hasFood()) {
+                org.bukkit.inventory.meta.components.FoodComponent foodComponent = contextBox.getMeta().getFood();
+                org.bukkit.inventory.ItemStack leftover = foodComponent.getUsingConvertsTo();
+                if (leftover != null) {
+                    org.bukkit.inventory.ItemStack newLeftover = ItemModifier.modifyItem(new ItemContextBox(contextBox.getViewer(), contextBox.getLocale(), ItemModifierContextType.EMPTY_NO_LORE, ItemModifierContext.EMPTY, leftover));
+                    if (newLeftover != null) {
+                        foodComponent.setUsingConvertsTo(newLeftover);
+                        contextBox.getMeta().setFood(foodComponent);
+                        modified = true;
+                    }
+                }
+            }
+        }
+
         if (!modified) return null;
 
         org.bukkit.inventory.ItemStack item = contextBox.getItem();
